File: apps/backend/app/modules/templates/services.py
from __future__ import annotations

import json
from typing import Any, Dict


def _json_size_bytes(d: Dict[str, Any]) -> int:
    try:
        return len(json.dumps(d, ensure_ascii=False).encode("utf-8"))
    except Exception:
        return 10**9


def _depth(d: Any) -> int:
    if not isinstance(d, dict):
        return 0
    if not d:
        return 1
    return 1 + max((_depth(v) for v in d.values()), default=0)


def validate_overlay(config: Dict[str, Any], limits: Dict[str, Any]) -> None:
    max_fields = int(limits.get("max_fields", 15))
    max_bytes = int(limits.get("max_bytes", 8192))
    max_depth = int(limits.get("max_depth", 2))

    # Count fields (shallow)
    field_count = 0
    def count_fields(obj: Any) -> None:
        nonlocal field_count
        if isinstance(obj, dict):
            field_count += len(obj)
            for v in obj.values():
                count_fields(v)
        elif isinstance(obj, list):
            for v in obj:
                count_fields(v)
    count_fields(config)

    if field_count > max_fields:
        raise ValueError(f"overlay_fields_exceeded:{field_count}>{max_fields}")
    if _json_size_bytes(config) > max_bytes:
        raise ValueError("overlay_bytes_exceeded")
    if _depth(config) > max_depth:
        raise ValueError("overlay_depth_exceeded")


def deep_merge(base: Dict[str, Any], over: Dict[str, Any]) -> Dict[str, Any]:
    out = dict(base)
    for k, v in over.items():
        if k in out and isinstance(out[k], dict) and isinstance(v, dict):
            out[k] = deep_merge(out[k], v)
        else:
            out[k] = v
    return out

File: apps/tenant/src/modules/templates/services.ts
import api from "../../services/api/client"

export interface UIConfig {
  [key: string]: any
}

export interface TemplatePackage {
  template_key: string
  version: string
  config: UIConfig
}

export interface TemplateOverlay {
  id?: string
  template_key?: string
  config: UIConfig
  active?: boolean
  created_at?: string
  updated_at?: string
}

export async function getUIConfig(): Promise<UIConfig> {
  return api.get('/api/v1/tenant/templates/ui-config').then(r => r.data)
}

export function deepMerge(base: Record<string, any>, over: Record<string, any>): Record<string, any> {
  const out = { ...base }
  for (const [k, v] of Object.entries(over)) {
    if (k in out && typeof out[k] === 'object' && typeof v === 'object' && !Array.isArray(out[k]) && !Array.isArray(v)) {
      out[k] = deepMerge(out[k], v)
    } else {
      out[k] = v
    }
  }
  return out
}

export function validateOverlay(
  config: Record<string, any>,
  limits: { max_fields?: number; max_bytes?: number; max_depth?: number } = {}
): { valid: boolean; error?: string } {
  const maxFields = limits.max_fields ?? 15
  const maxBytes = limits.max_bytes ?? 8192
  const maxDepth = limits.max_depth ?? 2

  let fieldCount = 0
  const countFields = (obj: any): void => {
    if (typeof obj === 'object' && obj !== null) {
      if (Array.isArray(obj)) {
        obj.forEach(countFields)
      } else {
        fieldCount += Object.keys(obj).length
        Object.values(obj).forEach(countFields)
      }
    }
  }

  countFields(config)

  if (fieldCount > maxFields) {
    return { valid: false, error: `overlay_fields_exceeded:${fieldCount}>${maxFields}` }
  }

  const bytes = new Blob([JSON.stringify(config)]).size
  if (bytes > maxBytes) {
    return { valid: false, error: 'overlay_bytes_exceeded' }
  }

  const getDepth = (obj: any): number => {
    if (typeof obj !== 'object' || obj === null) return 0
    if (Array.isArray(obj)) {
      return obj.length === 0 ? 1 : 1 + Math.max(...obj.map(getDepth))
    }
    const keys = Object.keys(obj)
    return keys.length === 0 ? 1 : 1 + Math.max(...keys.map(k => getDepth(obj[k])))
  }

  if (getDepth(config) > maxDepth) {
    return { valid: false, error: 'overlay_depth_exceeded' }
  }

  return { valid: true }
}
